alf/algorithms/algorithm.py

```diff
@@ -171,6 +185,96 @@ def __init__(self,
+        self._checkpoint_pre_loaded = False
+
+    def __init_subclass__(cls, *args, **kwargs):
+        """This function is called at the creation of sub-classes of this class.
+        Here we customize the ``__init__``function of the input ``cls`` to have
+        a post init call when the input ``cls`` is the sub-class.
+        """
+        super().__init_subclass__(*args, **kwargs)
+
+        # use ``functools.warps`` to keep the signature of the original
+        # ``__init__`` function, to ensure alf.config work correctly.
+        @functools.wraps(cls.__init__)
+        def new_init(self, *args, init=cls.__init__, **kwargs):
+            init(self, *args, **kwargs)
+            if cls is type(self):
+                self._post_init()
+
+        cls.__init__ = new_init
+
+    def _post_init(self):
+        """This function will be called automatically in the sub-class
+        at the end of the __init__ function, in order to activate _post_init
+        functionalities.
+        Algorithms can overwrite this function to provide customized post init
+        behaviors.
+        """
+        self._preload_checkpoint()
+
+    def _preload_checkpoint(self):
+        """Preload checkpoint to the algorithm, based on the specified ``checkpoint``.
+        """
+
+        def _remove_prefix(s, prefix):
+            if s.startswith(prefix):
+                return s[len(prefix):]
+            else:
+                return s
+
+        if self._checkpoint is not None:
+            try:
+
+                prefix_and_path = self._checkpoint.split('@')
+                assert len(prefix_and_path) in [
+                    1, 2
+                ], ("invalid checkpoint: "
+                    "{}").format(prefix_and_path)
+
+                if len(prefix_and_path) == 1:
+                    # only path is provided
+                    checkpoint_path = prefix_and_path[0]
+                    checkpoint_prefix = ''
+                else:
+                    # only path is provided
+                    checkpoint_prefix, checkpoint_path = prefix_and_path
+
+                assert (checkpoint_prefix == ''
+                        or 'alg.' in checkpoint_prefix), "wrong prefix"
+
+                map_location = None
+                if not torch.cuda.is_available():
+                    map_location = torch.device('cpu')
+
+                checkpoint = torch.load(
+                    checkpoint_path, map_location=map_location)['alg']
+
+                if checkpoint_prefix != '':
+                    # the case when the checkpoint is a subset of the full
+                    # checkpoint file filter
+                    prefix = _remove_prefix(checkpoint_prefix, 'alg.')
+                    checkpoint = {
+                        _remove_prefix(k, prefix + '.'): v
+                        for k, v in checkpoint.items() if k.startswith(prefix)
+                    }
+
+                self.load_state_dict(checkpoint, strict=True)
```

Comment:
This part can be an independent function. Loading a module by prefix from a checkpoint file may be useful in other context.

Reply:
Good suggestion. Extracted it as an independent function.